File: supermarket-pricing/csharp/src/SupermarketPricing/Weight.cs
namespace SupermarketPricing;

/// <summary>
/// A non-negative weight in kilograms. Rejects negative values at construction.
/// </summary>
public readonly record struct Weight
{
    public decimal Kg { get; }

    public Weight(decimal kg)
    {
        if (kg < 0)
        {
            throw new ArgumentOutOfRangeException(nameof(kg), "Weight must not be negative");
        }
        Kg = kg;
    }

    public static Weight Zero => new(0m);
}

File: supermarket-pricing/typescript/src/Weight.ts
/**
 * Non-negative weight in kilograms.
 */
export class Weight {
  readonly kg: number;

  constructor(kg: number) {
    if (kg < 0) {
      throw new RangeError('Weight must not be negative');
    }
    this.kg = kg;
  }

  static readonly zero = new Weight(0);
}
